src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -95,7 +95,7 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
 
   @inline private def checkWithinBounds(lo: Int, hi: Int) = {
     if (lo < 0) throw new IndexOutOfBoundsException(s"$lo is out of bounds (min 0, max ${size0-1})")
-    if (hi > size0) throw new IndexOutOfBoundsException(s"$hi is out of bounds (min 0, max ${size0 - 1})")
+    if (hi > size0) throw new IndexOutOfBoundsException(s"${hi - 1} is out of bounds (min 0, max ${size0 - 1})")
```

Comment:
The solution suggested in scala/bug#12176 uses `$lo` here, but after a bit consideration it is wrong because it breaks with a situation where performing operations targeted for a range of indexes like this:
https://github.com/scala/scala/pull/9249/files#diff-421aa7b6b8733d6090f2148be711352af11cc34b2e6978c84051d2ecf0561bd6R27

Use of `$lo` produces incorrect message of `2 is out of bounds (min 0, max 2)`.